nym-vpn-core/nym-vpn-lib/src/mixnet_processor.rs

```diff
@@ -158,115 +167,241 @@ impl MixnetProcessor {
-                                let msg = format!("Received info response from the mixnet: {}", info.reply);
-                                match info.level {
-                                    InfoLevel::Info => log::info!("{msg}"),
-                                    InfoLevel::Warn => log::warn!("{msg}"),
-                                    InfoLevel::Error => log::error!("{msg}"),
+                                // Consider not including packets that are ICMP ping replies to our beacon
+                                // in the responses. We are defensive here just in case we incorrectly
+                                // label real packets as ping replies to our beacon.
+                                if let Err(err) = self.tun_device_sink.send(packet.into()).await {
+                                    error!("Failed to send packet to tun device: {err}");
                                 }
                             }
-                        },
+                        }
+                        Ok(None) => {}
                         Err(err) => {
-                            // The exception to when we are not expecting a response, is when we
-                            // are sending a ping to ourselves.
-                            if let Ok(request) = IpPacketRequest::from_reconstructed_message(&reconstructed_message) {
-                                match request.data {
-                                    IpPacketRequestData::Ping(ref ping_request) if ping_request.reply_to == our_address => {
-                                        self.connection_event_tx
-                                            .unbounded_send(ConnectionStatusEvent::MixnetSelfPing)
-                                            .unwrap();
-                                    },
-                                    ref request => {
-                                        info!("Received unexpected request: {request:?}");
-                                    }
-                                }
-                            } else {
-                                warn!("Failed to deserialize reconstructed message: {err}");
-                            }
+                            error!("Mixnet listener: {err}");
                         }
                     }
                 }
+                else => {
+                    error!("Mixnet listener: mixnet stream ended");
+                    break;
+                }
             }
         }
-        debug!("MixnetProcessor: Exiting");
-        Ok(tun_device_sink
-            .reunite(tun_device_stream)
-            .expect("reunite should work because of same device split")
-            .into_inner())
+
+        debug!("Mixnet listener: Exiting");
+        self.tun_device_sink
+    }
+
+    fn start(self) -> JoinHandle<SplitSink<Framed<AsyncDevice, TunPacketCodec>, Vec<u8>>> {
+        tokio::spawn(async move { self.run().await })
```

Comment:
Since `tokio::spawn()` accepts anything that implements Future, I believe you could simply do the following: `tokio::spawn(self.run())`.

Reply:
On a side note, the call site can do the `tokio::spawn()` itself if it wants too. Otherwise it looks like we duplicate what `tokio::spawn()` already offers.